docs/proposals/promotion-strategy.md

```diff
@@ -0,0 +1,453 @@
+
+## Summary
+
+We propose a pattern and tooling to implement promotion and reversion based on common SCM primitives: Pull Requests and Commit Statuses. By representing the promotion/reversion process using these primitives, we keep the entire GitOps system "in git." We can take advantage of developers' existing comfort with git tooling as well as the vast ecosystem of tooling integrating with git and its supporting SCMs.
+
+### Goals
+
+* The user experience must be intuitive
+    * No confusion about the state of my change in the promotion process
+    * No waiting around for the system to communicate its state
+* There must be no intentional drift between hydrated branches and the cluster
+* The tool must be usable by non-Argo GitOps tools - e.g. Flux should be able to integrate with this system without using esoteric adapters
+
+### Non-Goals
+
+* This is not an image updater; automated tag bumps must be done by another system, and this system can promote that change
+
+## Proposal
+
+#### Promotion
+
+This proposal introduces a system of push-by-pr-merge. Instead of representing the change promotion state in some external system, it is represented as PRs against environment branches containing "hydrated" (or "rendered") manifests. The "wait" stages of the promotion process (performance tests, manual approvals, etc.) are represented as commit statuses blocking PR merges. The state of the entire system is visible in the git/SCM interface. This system is generalizable to many organizations, relying on near-ubiquitous SCM concepts, allowing the promotion tool to be open-sourced.
+
+![Diagram showing the promote-by-pr-merge strategy. Changes originate as PRs against the main (DRY) branch. A tool then generates PRs for each environment with the hydrated manifests. The tool merges the PRs in the designated order. Changes are immediately synced out to the cluster on PR merge.](images/promote-by-pr-merge.png)
+
+#### Reversion
+
+This proposal introduces a system for automated, atomic rollbacks via PRs. When a problem is detected, a PR is automatically opened against the affected environment's hydrated manifest branch reverting to the last known-healthy commit. The PR may be automatically merged, may wait for certain checks to pass, or may wait for a user to manually merge it.
+
+### Concepts
+
+#### Change
+
+Every change is represented as a commit on a "DRY" branch in a git repo. The contents of the branch depend completely on the manifest hydration tool. For example, Kustomize may include kustomization.yaml and a few related manifest files; Helm may include values.yaml and supporting chart files; and so on.
+
+#### Environment
+
```

Comment:
We generally advise against using branchers for environments, and instead recommend using folders. I would hate to enshrine branches in this way. I think you're probably familiar with the arguments for why that is but [here's a refresher](https://codefresh.io/blog/stop-using-branches-deploying-different-gitops-environments/).

Reply:
I need to clarify this section a bit... definitely agree "directories for environments" with respect to _dry_ manifests (e.g. your kustomize bases and overlays). But a separate branch for each environment's _hydrated_ manifests. So the user's "write" interface is directories. The "read" interface is branches.

It also has the nice side-benefit of letting you use SCMs' "compare" interface to compare environments' hydrated manifests. And you get a full history of the hydrated manifests for just one environment by looking at the branch's commits.

Reply:
From the blog post:

> If you follow a branch-based approach for your environments:
> 
> 1. You need to have 13 long living Git branches at all times.
> 2. You need 13 pull requests for promoting a single change across all environments.
> 3. You have a two dimensional promotion matrix with 5 steps upwards and 2-3 steps outwards.
> 4. The possibilities for wrong merges, configuration drift and ad-hoc changes is now non-trivial across all environment combinations.

I think these all apply to a system where the _user_ has to interact directly with those branches. The proposal would automate all those interactions.

Reply:
Exactly. Cheap and often has an awesome UI "for free."